nntrainer/tensor/half_tensor.cpp

```diff
@@ -284,8 +300,67 @@ TensorV2 &HalfTensor::apply(std::function<_FP16(_FP16)> f,
       for (unsigned int c = 0; c < channel(); ++c) {
         for (unsigned int h = 0; h < height(); ++h) {
           for (unsigned int w = 0; w < width(); ++w) {
-            output.setValue(b, c, h, w,
-                            f(((_FP16 *)getData())[getIndex(b, c, h, w)]));
+            output.setValue(b, c, h, w, f(getValue(b, c, h, w)));
+          }
+        }
+      }
+    }
+  }
+
+  return output;
+}
+
+TensorV2 HalfTensor::multiply_strided(TensorV2 const &m, TensorV2 &output,
+                                      const float beta) const {
+  CREATE_V2_IF_EMPTY_DIMS(output, dim, nullptr);
+
+  if (size() != m.size() || size() != output.size())
+    throw std::invalid_argument(
+      "Strided multiplication does not support broadcasting");
+
+  NNTR_THROW_IF(getData() == nullptr, std::invalid_argument)
+    << getName() << " is not allocated";
+  NNTR_THROW_IF(m.getData<_FP16>() == nullptr, std::invalid_argument)
+    << m.getName() << " is not allocated";
+  NNTR_THROW_IF(output.getData<_FP16>() == nullptr, std::invalid_argument)
+    << output.getName() << " is not allocated";
+
+  if (strides[3] != 1 || m.getStrides()[3] != 1 ||
+      output.getStrides()[3] != 1 || beta != 0.0) {
+    for (unsigned int b = 0; b < batch(); ++b) {
+      for (unsigned int c = 0; c < channel(); ++c) {
+        for (unsigned int h = 0; h < height(); ++h) {
+          for (unsigned int w = 0; w < width(); ++w) {
+            output.addValue(
+              b, c, h, w, getValue(b, c, h, w) * m.getValue<_FP16>(b, c, h, w),
```

Comment:
do we need _FP16 in ```m.getValue```? I thought we do not need to specify the type in TensorV2 level.

Reply:
We do. m is a TensorV2, not a HalfTensor, which means getValue is a Tensor::getValue that requires template argument.